File: apps/scraper/query/sources/sports.py
"""Sun Devil Athletics schedules and results for one sport."""

from __future__ import annotations

from scraper.core.types import QueryParam, QuerySource
from scraper.query.params import choices, code

# Path segment of each sport on thesundevils.com.
_SPORTS = {
    "football": "football",
    "men's basketball": "mens-basketball",
    "women's basketball": "womens-basketball",
    "baseball": "baseball",
    "softball": "softball",
    "volleyball": "womens-volleyball",
    "soccer": "womens-soccer",
    "hockey": "mens-ice-hockey",
    "wrestling": "wrestling",
}


def to_url(params: dict[str, str]) -> str:
    """The schedule page of the sport."""
    return f"https://thesundevils.com/sports/{code(params, 'sport', _SPORTS)}/schedule"


QUERY = QuerySource(
    key="sports",
    description="Fetch the Sun Devil schedule and recent results for one sport.",
    params=(QueryParam("sport", "Sport.", required=True, choices=choices(_SPORTS)),),
    to_url=to_url,
    needs_js=True,
)

File: apps/engine/src/runtime/tools/knowledge/search/sports.rs
//! The Sun Devil schedule and recent results for one sport.

use super::{LiveSource, Param};

/// Sun Devil Athletics schedules.
pub struct Sports;

const SPORTS: &[&str] = &[
    "football",
    "men's basketball",
    "women's basketball",
    "baseball",
    "softball",
    "volleyball",
    "soccer",
    "hockey",
    "wrestling",
];

const PARAMS: &[Param] = &[Param::one_of("sport", "Sport.", SPORTS).required()];

impl LiveSource for Sports {
    fn key(&self) -> &'static str {
        "sports"
    }

    fn hint(&self) -> &'static str {
        "Sun Devil schedules and results"
    }

    fn label(&self) -> &'static str {
        "Sun Devil Athletics"
    }

    fn category(&self) -> &'static str {
        "sports"
    }

    fn params(&self) -> &'static [Param] {
        PARAMS
    }
}
